openvdb/openvdb/tools/Morphology.h

```diff
@@ -63,540 +59,835 @@ enum NearestNeighbors { NN_FACE = 6, NN_FACE_EDGE = 18, NN_FACE_EDGE_VERTEX = 26
+                    const auto& leaf = mManager.leaf(idx);
+                    if (leaf.isEmpty()) continue;
+                    // original bit-mask of current leaf node
+                    MaskType& newMask = nodeMasks[idx];
+                    cache.erode(leaf, newMask);
+                }
+            });
+
+            // update the masks after all nodes have been eroded
+            tbb::parallel_for(range,
+                [&](const tbb::blocked_range<size_t>& r){
+                for (size_t idx = r.begin(); idx < r.end(); ++idx)
+                    mManager.leaf(idx).setValueMask(nodeMasks[idx]);
+            });
         }
-    private:
-        std::vector<MaskType>& mMasks;
-        ManagerType& mManager;
-        bool mSaveMasks;
-    };// MaskManager
-
-    struct UpdateMasks {
-        UpdateMasks(const std::vector<MaskType>& masks, ManagerType& manager)
-            : mMasks(masks), mManager(manager) {}
-        void update() { tbb::parallel_for(mManager.getRange(), *this); }
-        void operator()(const tbb::blocked_range<size_t>& r) const {
-            for (size_t i=r.begin(); i<r.end(); ++i) mManager.leaf(i).setValueMask(mMasks[i]);
+    }
+
+    // if prune, replace any inactive nodes
+    if (prune) {
+        tools::prune(mManager.tree(),
+            zeroVal<typename TreeType::ValueType>(),
+            this->getThreaded());
+        mManager.rebuild(!this->getThreaded());
+    }
+}
+
+template <typename TreeType>
+void Morphology<TreeType>::dilateVoxels(const size_t iter,
+    const NearestNeighbors nn,
+    const bool prune,
+    const bool preserveMaskLeafNodes)
+{
+    if (iter == 0) return;
+
+    const bool threaded = this->getThreaded();
+
+    // Actual dilation op. main implementation is single threaded. Note that this
+    // is templated (auto-ed) as the threaded implemenation may call this with a
+    // different value type to the source morphology class
+    // @note  GCC 6.4.0 crashes trying to compile this lambda with [&] capture
+    auto dilate = [iter, nn, threaded](auto& manager, const bool collapse) {
+
+        using LeafManagerT = typename std::decay<decltype(manager)>::type;
+        using TreeT = typename LeafManagerT::TreeType;
+        using ValueT = typename TreeT::ValueType;
+        using LeafT = typename TreeT::LeafNodeType;
+
+        // this is only used for the impl of copyMasks
+        Morphology<TreeT> m(manager);
+        m.setThreaded(threaded);
+
+        TreeT& tree = manager.tree();
+        tree::ValueAccessor<TreeT> accessor(tree);
+
+        // build cache objects
+        typename Morphology<TreeT>::NodeMaskOp cache(accessor, nn);
+        std::vector<MaskType> nodeMasks;
+        std::vector<std::unique_ptr<LeafT>> nodes;
+        const ValueT& bg = tree.background();
+        const bool steal = iter > 1;
+
+        for (size_t i = 0; i < iter; ++i) {
+            if (i > 0) manager.rebuild(!threaded);
```

Comment:
Did you try this without using a LeafManager - ie just a linear list of nodes with a on/off toggle for each one? Feels like you might be spending some time extracting and re-inserting leaf nodes and rebuilding the leaf manager with this technique?

Reply:
For iter>1 I always need to call `getNodes`. So unless `initLeafArray` is slower than `getNodes`, the leaf manager way is better.